util/process_perf_data.py

```diff
@@ -0,0 +1,204 @@
+CHECKPOINTS = [
+    { 'name': "Fresh start", 'path': "Iteration 1/Start" },
+    { 'name': "Fresh start [+30s]", 'path': "Iteration 1/StartSettled" },
+    { 'name': "After tabs open", 'path': "Iteration 5/TabsOpen" },
+    { 'name': "After tabs open [+30s]", 'path': "Iteration 5/TabsOpenSettled" },
+    { 'name': "After tabs open [+30s, forced GC]", 'path': "Iteration 5/TabsOpenForceGC" },
+    { 'name': "Tabs closed", 'path': "Iteration 5/TabsClosed" },
+    { 'name': "Tabs closed [+30s]", 'path': "Iteration 5/TabsClosedSettled" },
+    { 'name': "Tabs closed [+30s, forced GC]", 'path': "Iteration 5/TabsClosedForceGC" }
+]
+
+# A description of each perfherder suite and the path to its values.
+PERF_SUITES = [
+    { 'name': "Resident Memory", 'node': "resident" },
+    { 'name': "Explicit Memory", 'node': "explicit" },
+    { 'name': "Heap Unclassified", 'node': "explicit/heap-unclassified" },
+    { 'name': "JS", 'node': "js-main-runtime" },
+    { 'name': "Images", 'node': "explicit/images" }
+]
+
+def get_node_value(datapoint, nodes):
+    """
+    Retrieves a nested value from the nodes dictionary.
+
+    :param nodes: A dictionary of nested value nodes.
+    :param datapoint: A path to the desired value.
+    """
+    node = nodes
+    for branch in datapoint.split('/'):
+        if node and branch in node:
+            node = node[branch]
+        else:
+            raise KeyError("Datapoint not found: %s" % datapoint)
+
+    if type(node) in [int, long]:
+        return node
+    else:
+        return node.get('_val')
+
+
+def create_suite(name, node, data):
+    """
+    Creates a suite suitable for adding to a perfherder blob. Calculates the
+    geometric mean of the checkpoint values and adds that to the suite as
+    well.
+
+    :param name: The name of the suite.
+    :param node: The path of the data node to extract data from.
+    :param data: The dataset to retrieve data from.
+    """
+    suite = {
+        'name': name,
+        'subtests': [],
+        'lowerIsBetter': True
+    }
+
+    total = 0
+    for checkpoint in CHECKPOINTS:
+        subtest = {
+            'name': checkpoint['name'],
+            'value': get_node_value('/'.join([ checkpoint['path'], node ]), data),
+            'lowerIsBetter': True,
+            'units': 'bytes'
+        }
+        suite['subtests'].append(subtest);
+        total += math.log(subtest['value'])
+
+    # Add the geometric mean. For more details on the calculation see:
+    #   https://en.wikipedia.org/wiki/Geometric_mean#Relationship_with_arithmetic_mean_of_logarithms
+    suite['value'] = math.exp(total / len(CHECKPOINTS))
+
+    return suite
+
+
+def create_perf_data(nodes):
+    """
+    Builds up a performance data blob suitable for submitting to perfherder.
+    """
+    perf_blob = {
+        'framework': { 'name': 'awsy' },
+        'suites': []
+    }
+
+    for suite in PERF_SUITES:
+        perf_blob['suites'].append(create_suite(suite['name'], suite['node'], nodes))
```

Comment:
I decided to include the full set of data (so a value for each checkpoint) in all of the suites. The reason we don't do that on the AWSY site is because we merged the 3 values into one graph as "Miscellaneous," but in perfherder we don't have that limitation.